EXCERSICE 1 :
Load the built-in diabetes dataset (features → a number).
Split 80/20 with random_state=42.
fit a LinearRegression, predict on the test set.
Print RMSE, MAE and R². Compare to a "predict the mean" baseline.

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import (
    mean_squared_error,
    mean_absolute_error,
    r2_score
)

#Load the dataset 
datset = load_diabetes()
x = dataset.data
y = dataset.target

#Split dataset (80% train , 20 %test)

X_train , X_test , Y_train , Y_test = train_test_split(
    x,
    y,
    test_size = 0.2,
    random_state = 42 #splitting data or shuffling samples—happen in the exact same order every single run
)

#CREATE AND RUN MODEL:
model = LinearRegression()

model.fit(X_train,Y_train)

pred = model.predict(X_test)

rmse = np.sqrt(mean_squared_error(Y_test,pred))

mae = mean_absolute_error(Y_test , pred)

r2 = r2_score(y_test , pred)

print("Linear Regression")
print(f"RMSE : {rmse:.2f}")
print(f"MAE  : {mae:.2f}")
print(f"R²   : {r2:.4f}")


EXCERSICE 2 :
- Load `breast_cancer` (features → malignant/benign).
- Split with `stratify=y`, fit a `LogisticRegression`.
- Print the confusion matrix.
- Print accuracy, precision, recall, F1 — and read them out loud.

In [5]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

X ,y = load_breast_cancer(return_X_y = True) #features , labels

X_train , X_test , y_train , y_test = train_test_split(
    X , y , test_size = 0.2 , random_state=42 , stratify=y) #Stratify : Divide such that divides all features evenly
    
model = LogisticRegression(max_iter = 5000)
model.fit(X_train , y_train)
pred = model.predict(X_test)

from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test,pred)
print(accuracy)

#Print the confusion matrix.
conf = confusion_matrix(y_test , pred)
print(conf)

0.9649122807017544
[[39  3]
 [ 1 71]]


Exercise 3 · buffer / stretch
Prove that accuracy lies:
Build a fake imbalanced set: 980 negatives, 20 positives.
- Make a "model" that always predicts 0.
- Print its accuracy, recall and precision — 
watch accuracy say 0.98 while recall says 0.00.


In [30]:
#Build a fake imbalanced set: 980 negatives, 20 positives.
import numpy as np
import pandas as pd
n_neg = 980
n_pos = 20
np.random.seed(42)
neg_x1 = np.random.normal(loc=0, scale=1.5, size=n_neg)
neg_x2 = np.random.normal(loc=0, scale=1.5, size=n_neg)

pos_x1 = np.random.normal(loc=0.8, scale=1.5, size=n_pos)
pos_x2 = np.random.normal(loc=0.8, scale=1.5, size=n_pos)

#Two features: 
X1 = np.concatenate([neg_x1, pos_x1])
X2 = np.concatenate([neg_x2, pos_x2])
#Label:take the list [0] and repeat it 980
y = np.array([0]*n_neg + [1]*n_pos)
#[980 zeros]  +  [20 ones]  =  [980 zeros, then 20 ones]  (1000 items total)

df = pd.DataFrame({"feature_1": X1, "feature_2": X2, "label": y})
df = df.sample(frac=1, random_state=42).reset_index(drop=True)  # shuffle

df["label"].value_counts()

X = df[["feature_1","feature_2"]]
y = df["label"]
X.head()
y.head()

0    0
1    0
2    0
3    0
4    0
Name: label, dtype: int64

In [32]:
#Make a "model" that always predicts 0.
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score , precision_score , recall_score

X_train , X_test , y_train , y_test = train_test_split(
    X , y , test_size=0.2 ,random_state=42 , stratify=y
)
model = LogisticRegression(max_iter=5000)
model.fit(X_train , y_train)
pred = model.predict(X_test)

acc = accuracy_score(y_test , pred)
prec = precision_score(y_test , pred)
recall = recall_score(y_test , pred)
print(acc , prec , recall)
print(y_test.value_counts())

print(np.unique(pred, return_counts=True))

0.98 0.0 0.0
label
0    196
1      4
Name: count, dtype: int64
(array([0]), array([200]))


D:\Anaconda\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


We can fix this by using 
LogisticRegression(max_iter=5000, class_weight="balanced")

Homework:
Use wine data instead

In [13]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

X ,y = load_breast_cancer(return_X_y = True) #features , labels

X_train , X_test , y_train , y_test = train_test_split(
    X , y , test_size = 0.2 , random_state=42 , stratify=y) #Stratify : Divide such that divides all features evenly
    
model = LogisticRegression(max_iter = 5000)
model.fit(X_train , y_train)
pred = model.predict(X_test)

from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test,pred)
print(accuracy)

0.9649122807017544


#Quetion : Why does the accuracy doesnt increase after a certain point : 
Every model has a limit.

For example

Decision Tree       → 100%
Random Forest       → 99%
SVM                 → 99%
Logistic Regression → 96%
